lib/block.js

```diff
@@ -9,122 +9,122 @@ export default class Block {
     payload.write(this.parent.replace('0x', ''), 0, 'hex');
-    const big = ~~(this.height / MAX_UINT32);  // eslint-disable-line no-bitwise
-    payload.writeUInt32BE(big, 32);
-    const low = (this.height % MAX_UINT32) - big;
-    payload.writeUInt32BE(low, 36);
+    writeUint64(payload, this.height, 32);
     payload.write(this.merkleRoot().replace('0x', ''), 40, 'hex');
     return payload;
   }
 
   sigHash() {
-    return ethUtil.sha3(this.unsigedHeader(Buffer.alloc(72, 0)));
+    return ethUtil.sha3(this.unsignedHeader(Buffer.alloc(72, 0)));
   }
 
   sign(privKey) {
-    const privBuf = new Buffer(privKey.replace('0x', ''), 'hex');
+    const privBuf = toBuffer(privKey);
     const sig = ethUtil.ecsign(this.sigHash(), privBuf);
     this.v = sig.v;
     this.r = sig.r;
     this.s = sig.s;
-    return [sig.v, `0x${sig.r.toString('hex')}`, `0x${sig.s.toString('hex')}`];
+    return [sig.v, bufferToHex(sig.r), bufferToHex(sig.s)];
   }
 
   hash() {
     if (!this.v || !this.r || !this.s) {
       throw new Error('not signed yet');
     }
-    const payload = this.unsigedHeader(Buffer.alloc(137, 0));
+    const payload = this.unsignedHeader(Buffer.alloc(137, 0));
 
     payload.writeUInt8(this.v, 72);
     this.r.copy(payload, 73);
     this.s.copy(payload, 105);
     return ethUtil.bufferToHex(ethUtil.sha3(payload));
   }
 
-  //  _txData = [ 32b blockHash, 32b r, 32b s, (4b offset, 8b pos, 1b v, ..00.., 1b txData),
-  // 32b txData, 32b proof, 32b proof ]
-  proof(txData, pos, proof) {
+  // Returns merkle proof for given tx
+  // [
+  // 32b - blockHash
+  // 32b - r
+  // 32b - s,
+  // 32b - (4b tx length, 4b pos, 1b v, ..00.., Xb txData),
+  // 32b - txData,
+  // ... <more 32b tx data if needed>
+  // 32b - proof,
+  // ... <more 32b proofs if needed>
+  // ]
+  proof(tx) {
+    const txData = tx.toRaw();
+    const pos = this.txHashList.indexOf(tx.hash());
+    if (pos < 0) {
+      throw Error('tx not in the block');
+    }
     const slices = [];
     slices.push(this.hash());
-    slices.push(`0x${this.r.toString('hex')}`);
-    slices.push(`0x${this.s.toString('hex')}`);
-
-    const sliceCount = Math.floor(txData.length / 32) + 1;
-
-    let payload = Buffer.alloc(32);
+    slices.push(bufferToHex(this.r));
+    slices.push(bufferToHex(this.s));
+
+    // Slices with tx metadata, v-sig and tx data
+    // metadata size: 4 bytes tx length + 4 bytes tx index + 1 byte v-sig = 9 bytes
+    const metadataAndTxLength = 9 + txData.length;
+    // align to 32 bytes, so that e.g. 38 bytes will be stored in 64 bytes buffer
+    const doubleWordAlignedLength = metadataAndTxLength + (32 - (metadataAndTxLength % 32));
+    const txPayload = Buffer.alloc(doubleWordAlignedLength);
     // tx length
-    payload.writeUInt32BE(txData.length, 0);
-    // offset
-    let offset = ((txData.length % 32) <= 19) ? 32 - (txData.length % 32) : (32 - (txData.length % 32)) + 32;
-    offset += 96;
-    payload.writeUInt8(offset, 0);
```

Comment:
@johannbarbie this writes to offset 0 and txData length also written to offset 0. Was it an error?

Reply:
we use only 2 bytes of the length, so we can overwrite the first 2 bytes.

Reply:
stupid question.. the answer is in the code